adithyasolai/testudo-bank
# UI and UX improvements with Spring Security and Bootstrap

### Overview of changes made in this PR
This PR improves the user interface presented to the user and the user experience in multiple ways.

Bootstrap is added (using a JSP tag template for consistency across pages) for basic styling to make the website more visually appealing and responsive.

Spring Security is added to allow a user to make multiple transactions, only having to "log in" once.

Integration tests are slightly modified to support testing the application when Spring Security is present

### Preview of new UI

#### Login page
<img width="1490" alt="image" src="https://user-images.githubusercontent.com/38168072/166526168-e78d8063-b956-49c6-92b3-0ff3c3ebb680.png">

#### Account overview page
<img width="1498" alt="image" src="https://user-images.githubusercontent.com/38168072/166527653-4b0a66e5-321a-4ffa-b7af-3db17efad762.png">

#### Buy Crypto Page
<img width="1498" alt="image" src="https://user-images.githubusercontent.com/38168072/166528332-9559c4e9-53f6-4639-9b7d-fa9b37095ef8.png">



### Breakdown of changes

`pom.xml`
- Add Spring Security dependencies
- Add `jstl` dependency to enable additional template features for JSP

`python-sql-scripts/addCustomers.py`, `src/test/resources/clearDB.sql`, `src/test/resources/createDB.sql`
- Database schema modified to [support Spring Security](https://docs.spring.io/spring-security/reference/servlet/appendix/database-schema.html)

`src/main/webapp/WEB-INF/**`
- Modified JSP views extensively to improve UI

`src/main/java/net/testudobank/TransactionHistoryEntry.java`, `src/main/java/net/testudobank/TransferHistoryEntry.java`, `src/main/java/net/testudobank/OverdraftHistoryEntry.java`, `src/main/java/net/testudobank/CryptoTransactionHistoryEntry.java`
- Added models for transaction history to allow displaying transaction history in a table in JSP

`src/main/java/net/testudobank/CryptoPriceClient.java`
- Fixed bug causing cache miss for crypto price and improved method name 

`src/test/java/net/testudobank/tests/CryptoPriceClientTest.java`
- Added tests to test cache of price

`src/main/java/net/testudobank/MvcController.java`
- Modify controller methods to get user info from Spring Security context (`SecurityContextHolder.getContext().getAuthentication()`) rather than forgeable User model attribute. 
  - Username and password verification (and session info) is handled by Spring Security, not the controller
- Extract logic of deposit and withdraw to private methods
  - Other controller methods previously called the deposit and withdraw mappings directly. This is not proper and no longer worked in the case of transfer since the user id is from the security context.
  - Modify `updateAccountInfo` to populate new models for transaction entries to allow JSP to generate tables for transaction history
  - Add helpers `penniesToDollarString` and `dollarsToDollarString` for displaying currency amounts in a user-readable format (with dollar sign, commas, etc.)
 
`src/main/java/net/testudobank/TestudoBankRepository.java`
- Remove `getCustomerPassword` method. This is handled by Spring Security

`src/main/java/net/testudobank/User.java`
- Remove password field, not used anymore
- Change balance and price fields to strings to better display in UI
- Change transaction fields to appropriate model type to better display in UI

`src/main/java/net/testudobank/WebSecurityConfig.java`
- Add basic Spring Security Configuration to configure username and passwords stored in database
   - Use plaintext password encoder which **is not secure**, but simple for education purposes

`src/test/java/net/testudobank/helpers/MvcControllerIntegTestHelpers.java`
- Remove password from parameters, no longer used

`src/test/java/net/testudobank/tests/MvcControllerIntegTest.java`
- Add `@WithMockUser(username = "123456789")` to each test so the Spring Security context is populated with the correct user during test execution.
- Add support for testing with `MockMvc`
  - The current way of testing (calling controller methods directly) bypasses parts of Spring the requests hits before it gets to the controller. `MockMvc` tests more parts of the application
  - Add test case for accessing endpoints while not logged in using `MockMvc`
  - Modify `CryptoTransactionTester` to call endpoints using `MockMvc` for more substantial testing
  - Add `DynamicPropertySource` to configure database to that from Testcontainers
 - Add tests for multiple crypto transactions
